(PUP-8950) Handle SSL connection errors in new client

justinstoller opened the pull request:
Prior to this, despite using the existing validator, we did not handle
erros that might be thrown from validation. This refactors existing
validation handling to be used by our new client.

Magisus commented on lib/puppet/network/http/connection.rb:
```diff
@@ -313,14 +313,18 @@ def with_connection(site, &block)
+      self.class.handle_connection_error!(error, @verify, site.host)
+    end
+
+    def self.handle_connection_error!(error, verifier, host)
```
I could be wrong, but I feel like this shouldn't have the exclamation point here. In Ruby that usually is used to indicate a destructive/mutating operation on a collection or object.

Magisus commented on spec/unit/rest/client_spec.rb:
```diff
@@ -2,8 +2,61 @@
 describe Puppet::Rest::Client do
+  # Follows closely with spec/unit/network/http/connection_spec's
+  # 'ssl verifier' shared context
+  shared_context 'connection error handling' do
```
Can we use `shared_examples` here instead to make it clearer that we are sharing test cases as well as setup code?

Magisus commented:
I'm going to move the ticket to PUP, to make Travis happy.

justinstoller commented:
(;

I'm not sure why AppVeyor was failing. I'll guess I'll look at it if it fails again.

NB: I added moving the rest client loading to the top of the file even though it's no longer an issue with the new construction. I figured the comment was valuable enough for its own commit.

I dropped the other change (handling the inspect output) mentioned in the ticket since it isn't surfacing in the recent builds and there's an existing ticket to improve our error handling that I didn't want to conflict with.

Magisus commented:
My main issue with it is the namespacing; it would be nice if we had a more central location for the error handling logic. Maybe there is somewhere under `lib/util` we could put it? Perhaps in `Puppet::Util::SSL`, or even a new module. But I'm not that worried about it. Other than that, I'm fine with this.

Magisus commented:
Jenkins, please test this

Magisus commented:
Jenkins please test this

justinstoller commented:
Interested if anyone else has any opinions on if (and where) I should move this error handler.

If I don't here from anyone else by end of the day I'll move it into `Puppet::Util::SSL`.

Magisus commented:
jenkins please test this

puppetcla commented:
CLA signed by all contributors.

tvpartytonight commented on lib/puppet/util/ssl.rb:
```diff
@@ -50,4 +52,41 @@ def self.cn_from_subject(subject)
   end
+
+  ##
+  # Extract and format meaningful error messages from OpenSSS::OpenSSLErrors
```
OpenSSS 🐍  typo.

Magisus replied:
I fixed this in my PR that I'm working on.

tvpartytonight commented on lib/puppet/rest/client.rb:
```diff
@@ -31,43 +32,50 @@ def initialize(ssl_context:,
 
     # Make a PUT request to the specified URL with the specified params.
-    # @param [String] url the full path to query
+    # @param [URI] url the full path to query
```
I think this could be further specified to be `URI::HTTPS`; at least, that's the only thing we feed it from `Puppet::Rest::Route`.

Magisus replied:
I updated this in my PR that I'm working on.
